pkg/kubelet/images/image_pull_manager.go

```diff
@@ -0,0 +1,719 @@
+// ImagePullSecret objects.
+func imagePullSecretLess(a, b kubeletconfiginternal.ImagePullSecret) int {
+	if a.Namespace < b.Namespace {
+		return -1
+	} else if a.Namespace > b.Namespace {
+		return 1
+	}
+
+	if a.Name < b.Name {
+		return -1
+	} else if a.Name > b.Name {
+		return 1
+	}
+
+	if a.UID < b.UID {
+		return -1
+	}
+	return 1
+}
+
+// writeFile writes `content` to the file with name `filename` in directory `dir`.
+// It assures write atomicity by creating a temporary file first and only after
+// a successful write, it move the temp file in place of the target.
+func writeFile(dir, filename string, content []byte) error {
+	// create target folder if it does not exists yet
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return fmt.Errorf("failed to create directory %q: %w", dir, err)
+	}
+
+	targetPath := filepath.Join(dir, filename)
+	tmpPath := targetPath + ".tmp"
+	if err := os.WriteFile(tmpPath, content, 0600); err != nil {
+		return fmt.Errorf("failed to create temporary file %q: %w", tmpPath, err)
+	}
+
+	return os.Rename(tmpPath, targetPath)
+}
+
+// trimImageTagDigest removes the tag and digest from an image name
+func trimImageTagDigest(containerImage string) (string, error) {
+	imageName, _, _, err := parsers.ParseImageName(containerImage)
+	return imageName, err
+}
+
+func processDirFiles(dirName string, fileAction func(filePath string, fileContent []byte) error) error {
+	return filepath.WalkDir(dirName, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if path == dirName {
+			return nil
+		}
+
+		if d.IsDir() {
+			klog.V(4).InfoS("path is a directory, skipping", "path", path)
+			return nil
+		}
+
+		fileContent, err := os.ReadFile(path)
+		if err != nil {
+			klog.ErrorS(err, "skipping file, failed to read", "filePath", path)
+			return nil
+		}
+
+		if err := fileAction(path, fileContent); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func decodeIntent(d runtime.Decoder, objBytes []byte) (*kubeletconfiginternal.ImagePullIntent, error) {
+	obj, gvk, err := d.Decode(objBytes, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if gvk.Kind != "ImagePullIntent" ||
+		gvk.Group != kubeletconfigv1alpha1.SchemeGroupVersion.Group {
+		return nil, fmt.Errorf("unknown gvk for %s: %v", "ImagePullIntent", gvk)
+	}
+
+	intentObj, ok := obj.(*kubeletconfiginternal.ImagePullIntent)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert object to *ImagePullIntent: %v", obj)
+	}
+
+	return intentObj, nil
+}
+
+func decodePulledRecord(d runtime.Decoder, objBytes []byte) (*kubeletconfiginternal.ImagePulledRecord, error) {
+	obj, gvk, err := d.Decode(objBytes, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if gvk.Kind != "ImagePulledRecord" ||
+		gvk.Group != kubeletconfigv1alpha1.SchemeGroupVersion.Group {
+		return nil, fmt.Errorf("unknown gvk for %s: %v", "ImagePulledRecord", gvk)
+	}
```

Comment:
I think we can actually drop the gvk check if we're type-asserting *kubeletconfiginternal.ImagePulledRecord below